### DISCUSSION TITLE
Commercial solver options

### DISCUSSION THREAD

**Discussion by PeterKlein11** [blue-marble/gridpath#1154] [Q&A] [answered]
Can we use any solver that is compatible with Pyomo? We are looking at moving from Xpress (which works fine) to COPT. I see that COPT has a Pyomo example in their documentation. 

Thanks!

**Comment by anamileva** [comment]
Yes, you should be able to use any solver supported by Pyomo with GridPath. You can either use the `--solver` argument of `gridpath_run` or `gridpath_run_e2e`, or specify a solver and solver settings via the `solver_scenario_id` / `solver_options.csv`.

We don't have access to all Pyomo-supported solvers, so can't test them all and sometimes small tweaks may be required. I don't expect it will happen with COPT, but if you run into a problem, please raise a GitHub issue and we will try to help.

**Reply by PeterKlein11** [reply, marked as answer]
Some feedback on this - yes it is perfectly possible to use COPT to solve a model. Performance is very good and pricing very competitive. 

It requires copying the python file that comes with their installation "copt_pyomo.py" into the gridpath folder and adding
from copt_pyomo import * to the top of common_functions.py

The solver name must be specified as copt_direct. Happy to submit a pull request to check if coptpy is installed in the environment and then import it as a solver option if this is useful. Otherwise would it be possible to add quickly on the gridpath team's side?

**Reply by anamileva** [reply]
Thanks, @PeterKlein11! A PR for this functionality would be very welcome. It sounds very useful and I may build on it for other solvers also.
